modules/drivers/gnss/stream/tcp_stream.cc

```diff
@@ -0,0 +1,352 @@
+/******************************************************************************
+ * Copyright 2017 The Apollo Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *****************************************************************************/
+
+#include <arpa/inet.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <inttypes.h>
+#include <netinet/in.h>
+#include <netinet/tcp.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include <iostream>
+
+#include "ros/include/ros/ros.h"
+
+#include "modules/drivers/gnss/stream/stream.h"
+#include "modules/drivers/gnss/stream/tcp_stream.h"
+
+namespace apollo {
+namespace drivers {
+namespace gnss {
+
+TcpStream::TcpStream(const char* address, uint16_t port, uint32_t timeout_usec,
+                     bool auto_reconnect)
+    : sockfd_(-1), errno_(0), auto_reconnect_(auto_reconnect) {
+  peer_addr_ = inet_addr(address);
+  peer_port_ = htons(port);
+  timeout_usec_ = timeout_usec;
+}
+
+TcpStream::~TcpStream() { this->close(); }
+
+void TcpStream::open() {
+  int fd = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
```

Comment:
Just a FYI that later we may be able to merge some of the code here with the ones in canbus driver